### DISCUSSION TITLE
Using a keyboard mapper

### DISCUSSION THREAD

**Discussion by return215** [waydroid/waydroid#759] [Q&A]
I would like to use a tool to map my keyboard actions to touch controls. The touch controls should support tap, swipe, and multiple touch points. I mainly use it for games but that doesn't mean it can't be used for general purposes.

This was initially brought up in issue #466 with no response to this day. I was wondering of anyone have a suggestion to make it work, whether it is from Wayland itself or using a third-party app.

The app that I have been using is [KeyMapper](https://github.com/keymapperorg/KeyMapper). However, there are a few things that makes me not at ease. First, it does not support mapping to swiping gestures, which I need for certain games. Second, the developer is taking a break from maintaining the app, making me worry of potential breakage in the future.

**Comment by Blezz-tech** [comment]
please add this

**Comment by ayasa520** [comment]
I'm developing a keymapper  based on scrcpy, which provides Mumu-like keys. It doesn't integrate with waydroid, instead it overlays a transparent window to capture events. This is the simplest implementation I can manage

https://github.com/user-attachments/assets/1a4b2ec9-d153-4fb8-bb9f-f2a9eb976b4d

**Reply by jimed-rand** [reply]
Is that good to use it ? How did you that ?

**Reply by ayasa520** [reply]
> Is that good to use it ? How did you that ?

Here's the code.https://github.com/ayasa520/waydroid-helper/tree/controller

I'm working on refactoring the code, so I'm afraid it will take a while before I release it.

**Reply by jimed-rand** [reply]
Make it private when you afraid. And release it when it ready to publish.

**Reply by ayasa520** [reply]
> Make it private when you afraid. And release it when it ready to publish.

Then you should't ask🤷‍♀️

**Reply by 20univers08** [reply]
> > Make it private when you afraid. And release it when it ready to publish.
> 
> Then you should't ask🤷‍♀️

any progress :)

**Reply by ayasa520** [reply]
> any progress :)

https://github.com/waydroid-helper/waydroid-helper/blob/main/docs%2FKEY_MAPPING.md

**Reply by 20univers08** [reply]
What should I do ? Install waydroid-helper ?
Is that part of it ?

**Reply by ayasa520** [reply]
> What should I do ? Install waydroid-helper ? Is that part of it ?

yes it is

**Reply by an-abnormality** [reply]
Hey, I cannot get this to work. Hitting the editor just minimizes the Waydroid helper window, but doesn't seem to open anything

**Reply by ayasa520** [reply]
> Hey, I cannot get this to work. Hitting the editor just minimizes the Waydroid helper window, but doesn't seem to open anything

@an-abnormality  Please go to https://github.com/waydroid-helper/waydroid-helper/issues to submit an issue. I think it's most likely a bug related to window transparency. I have indeed encountered before on a specific distro (ubuntu22.04?) where windows would not appear when transparency was set to 0, but it worked normally when changed to 0.01. 

Could you try running `sudo sed -i 's/rgba(0, 0, 0, 0)/rgba(0, 0, 0, 0.01)/g' /usr/share/waydroid-helper/waydroid_helper/controller/ui/styles.py` and let me know if it works now?

**Reply by an-abnormality** [reply]
> > Hey, I cannot get this to work. Hitting the editor just minimizes the Waydroid helper window, but doesn't seem to open anything
> 
> @an-abnormality Please go to https://github.com/waydroid-helper/waydroid-helper/issues to submit an issue. I think it's most likely a bug related to window transparency. I have indeed encountered before on a specific distro (ubuntu22.04?) where windows would not appear when transparency was set to 0, but it worked normally when changed to 0.01.
> 
> Could you try running `sudo sed -i 's/rgba(0, 0, 0, 0)/rgba(0, 0, 0, 0.01)/g' /usr/share/waydroid-helper/waydroid_helper/controller/ui/styles.py` and let me know if it works now?

Hey, I got it to work. It seems like it was likely a GNOME issue. I switched from Fedora 42 Workstation to Fedora 42 KDE and the issue is no longer there.

**Comment by MRDGH2821** [comment]
Is there any update? 😅

**Comment by jimed-rand** [comment]
Anyway If you want that, why just using an XtMapper instead ? Waydroid are similarity to Anbox which just run an Android environment in Linux, not designed for gaming like another proprietary emulator for Windows like BlueStacks, Nox, or more.

If you want to use Waydroid for gaming environment, build by yourself with using XtMapper and if you're using Laptops, I recommend to use external keyboard and mouse instead.

**Comment by ayasa520** [comment]
https://github.com/user-attachments/assets/f42f8252-a09c-432e-995f-a663f8d5ea45

Inspired by [xtmapper](https://github.com/Xtr126/cage-xtmapper/), the [waydroid-helper](https://github.com/waydroid-helper/waydroid-helper) keymapper now works with cage, making it easier to run Waydroid in a Wayland kiosk environment. The keymapper works, but latency is a bit higher compared to xtmapper.

**Reply by unknown20243** [reply]
i know my question is stupid and i am wasting your time but please, please tell me how to save/apply and exit from waydroid helper keymapping

**Reply by ayasa520** [reply]
> i know my question is stupid and i am wasting your time but please, please tell me how to save/apply and exit from waydroid helper keymapping

https://github.com/waydroid-helper/waydroid-helper/blob/main/docs/KEY_MAPPING.md#getting-started
